pandajedi/jedibrokerage/AtlasAnalJobBroker.py

```diff
@@ -1665,43 +1683,160 @@ def doBrokerage(self, taskSpec, cloudName, inputChunk, taskParamMap):
-        ## compute new basic weight
+        ## compute new basic weight; only for User Analysis
         try:
-            weight_comparison_avail_sites = set(basic_weight_comparison_map.keys())
-            trr_sum = 0
-            nq_sum = 0
             n_avail_sites = len(basic_weight_comparison_map)
-            for vv in basic_weight_comparison_map.values():
-                trr_sum += vv['trr']
-                nq_sum += vv['nq']
             if n_avail_sites == 0:
                 tmpLog.debug('WEIGHT-COMPAR: zero available sites, skip')
-            elif trr_sum == 0:
-                tmpLog.debug('WEIGHT-COMPAR: zero sum of to-running-rate, skip')
             else:
-                _found_weights = False
-                while not _found_weights:
-                    trr_sum_avail = 0
-                    nq_sum_avail = 0
-                    n_avail_sites = len(weight_comparison_avail_sites)
-                    if n_avail_sites == 0:
-                        break
-                    for site in weight_comparison_avail_sites:
-                        vv = basic_weight_comparison_map[site]
-                        trr_sum_avail += vv['trr']
-                        nq_sum_avail += vv['nq']
-                    if trr_sum_avail == 0:
-                        break
-                    _found_weights = True
-                    for site in list(weight_comparison_avail_sites):
-                        vv = basic_weight_comparison_map[site]
-                        new_basic_weight = (vv['trr']/trr_sum_avail)*(25 + nq_sum_avail - n_avail_sites/2.0) - vv['nq'] + 1/2.0
-                        if new_basic_weight < 0:
-                            vv['new'] = 0
-                            weight_comparison_avail_sites.discard(site)
-                            _found_weights = False
-                        else:
-                            vv['new'] = new_basic_weight
+                # task class value
+                task_class_value = task_eval_dict.get('class', 1) if task_eval_dict is not None else 1
+                # get nFilesPerJob
+                if not inputChunk.isMerging:
+                    nFilesPerJob = taskSpec.getNumFilesPerJob()
+                else:
+                    nFilesPerJob = taskSpec.getNumFilesPerMergeJob()
+                if nFilesPerJob is None or nFilesPerJob < 1:
+                    nFilesPerJob = 1
+                # total unused input files, including psuedo inputs
+                indexVal = inputChunk.datasetMap[inputChunk.masterIndexName]
+                n_total_input_files = max(len(indexVal['datasetSpec'].Files) - indexVal['used'], 0)
+                # n_total_input_files = 0
+                # for datasetSpec in inputChunk.getDatasets(includePseudo=True):
+                #     n_total_input_files += len(datasetSpec.Files)
+                # estimate number of jobs to submit
+                n_jobs_to_submit = math.ceil(n_total_input_files/nFilesPerJob)
```

Comment:
@tmaeno 
The new weight algorithm is here and the dry-run results are dumped in the log.
The main issue now is, sometimes the estimated n_jobs_to_submit is far larger than real n_jobs_to_submit, making the entire algorithm less meaningful.
Do you know how to estimate number of jobs to submit in a brokerage cycle better?
Thanks.

Reply:
Hi @mightqxc 
I suppose that you have to create sub-chunks from the inputChunk like

https://github.com/PanDAWMS/panda-jedi/blob/master/pandajedi/jedicore/JediDBProxy.py#L1093

and count the number of sub-chunks that corresponds to the number of jobs, and then reset the internal counters using

https://github.com/PanDAWMS/panda-jedi/blob/master/pandajedi/jedicore/InputChunk.py#L132

The number of sub-chunks may depend on each site spec, such as maxwdir, but I think that you cound
ignore the difference.